### PR TITLE
[Workspace] Refactor workspace detail page

### PR DESCRIPTION
**Pull request by yubonluo**
### Description

Refactor workspace detail page

### Issues Resolved
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/7597

## Screenshot

https://github.com/user-attachments/assets/f5500124-4219-459f-9e72-46f66932825b

## Changelog
- feat: [Workspace] Refactor workspace detail page

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/workspace/public/workspace_client.ts [hunk 1]
```diff
@@ -232,6 +232,9 @@ export class WorkspaceClient {
     const result = await this.safeFetch<null>(this.getPath(id), { method: 'DELETE' });
 
     if (result.success) {
+      // After deleting workspace, need to reset current workspace ID.
+      this.workspaces.currentWorkspaceId$.next('');
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/workspace_client.ts:236]: the updateWorkspaceList will change the currentWorkspaceId$ I think.

File: src/plugins/workspace/public/components/workspace_form/workspace_detail_form.tsx [hunk 2]
```diff
@@ -66,79 +99,172 @@ export const WorkspaceDetailForm = (props: WorkspaceFormProps) => {
     handleDescriptionChange,
   } = useWorkspaceForm(props);
 
-  const isDashboardAdmin = application?.capabilities?.dashboards?.isDashboardAdmin ?? false;
+  useEffect(() => {
+    getNumberOfChanges(numberOfChanges);
+    getResetFunction(handleResetForm);
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_form/workspace_detail_form.tsx:104]: I think the reset function should better come from `useWorkspaceForm` hook?

File: src/plugins/workspace/public/components/workspace_form/workspace_detail_form.tsx [hunk 3]
```diff
@@ -66,79 +99,172 @@ export const WorkspaceDetailForm = (props: WorkspaceFormProps) => {
     handleDescriptionChange,
   } = useWorkspaceForm(props);
 
-  const isDashboardAdmin = application?.capabilities?.dashboards?.isDashboardAdmin ?? false;
+  useEffect(() => {
+    getNumberOfChanges(numberOfChanges);
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_form/workspace_detail_form.tsx:103]: getNumberOfChanges seems more like an util function, we do not need to pass it through props.

File: src/plugins/workspace/public/components/workspace_form/workspace_detail_form.tsx [hunk 4]
```diff
@@ -4,83 +4,136 @@
...
+  } = useWorkspaceFormContext();
   const disabledUserOrGroupInputIdsRef = useRef(
     defaultValues?.permissionSettings?.map((item) => item.id) ?? []
   );
 
+  const [isSaving, setIsSaving] = useState(false);
+
+  // Handle beforeunload event
+  useEffect(() => {
+    const handleBeforeUnload = (event: any) => {
```
Comment (ruanyl) [inline src/plugins/workspace/public/components/workspace_form/workspace_detail_form.tsx:76]: Can we avoid `any`?

File: src/plugins/workspace/public/components/workspace_form/workspace_detail_form_details.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    formErrors,
+    handleColorChange,
+    handleUseCaseChange,
+    handleNameInputChange,
+    handleDescriptionChange,
+  } = useWorkspaceFormContext();
+  const {
+    services: { workspaces },
+  } = useOpenSearchDashboards<CoreStart>();
+  const [value, setValue] = useState(formData.useCase);
```
Comment (ruanyl) [inline src/plugins/workspace/public/components/workspace_form/workspace_detail_form_details.tsx:54]: Why should it store `formData.useCase` as component local state? Cann't it just rely on `formData.useCase`?

File: src/plugins/workspace/public/application.tsx [hunk 6]
```diff
@@ -70,7 +72,14 @@ export const renderDetailApp = (
 ) => {
   ReactDOM.render(
     <OpenSearchDashboardsContextProvider services={services}>
-      <WorkspaceDetailApp {...props} />
+      <Router>
+        <Switch>
+          <Route>
+            <WorkspaceDetailApp {...props} />
+          </Route>
+          <Redirect to={`?tab=${DetailTab.Details}`} />
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/workspace/public/application.tsx:80]: I will delete the useless Redirect.

File: src/plugins/workspace/public/components/workspace_detail/association_data_source_modal.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+      setAllDataSources(filteredDataSources);
+      setOptions(
+        filteredDataSources.map((dataSource) => {
+          return { label: dataSource.title, id: dataSource.id };
+        })
+      );
+    });
+  }, [assignedDataSources, savedObjects]);
+
+  useEffect(() => {
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/workspace/public/components/workspace_detail/association_data_source_modal.tsx:54]: Sure, I will replace  `useEffect` and `useState` with `useMemo`.

File: src/plugins/workspace/public/components/workspace_detail/select_data_source_panel.tsx [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+          ) : (
+            <EuiText color="subdued">
+              <FormattedMessage
+                id="workspace.detail.dataSources.noAssociation.noAdmin.message"
+                defaultMessage="Contact your administrator to associate data sources with the workspace."
+              />
+            </EuiText>
+          )}
+        </EuiTextAlign>
+      )}
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/workspace/public/components/workspace_detail/select_data_source_panel.tsx:161]: Sure, I will refactor the code

File: src/plugins/workspace/public/components/workspace_detail/workspace_detail_panel.tsx [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+}
+export const WorkspaceDetailPanel = ({
+  useCaseUrl,
+  currentUseCase,
+  handleBadgeClick,
+  currentWorkspace,
+}: WorkspaceDetailPanelProps) => {
+  const owners = getOwners(currentWorkspace);
+  const formatDate = (lastUpdatedTime: string) => {
+    return moment(lastUpdatedTime).format('MMM DD, YYYY @ HH:mm:ss.SSS');
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/workspace/public/components/workspace_detail/workspace_detail_panel.tsx:66]: Sure, I will get the dateFormat from uiSettings

File: src/plugins/workspace/public/components/workspace_form/workspace_detail_form.tsx (hunk 4 above)
Comment (yubonluo) [reply src/plugins/workspace/public/components/workspace_form/workspace_detail_form.tsx:76]: Sure, I will update the type

File: src/plugins/workspace/public/components/workspace_form/workspace_detail_form_details.tsx (hunk 5 above)
Comment (yubonluo) [reply src/plugins/workspace/public/components/workspace_form/workspace_detail_form_details.tsx:54]: The aim is to reset the useCase select table value when user modify the useCase value and click the discard changes button.

**Review by ruanyl — APPROVED** [review]

**Review by SuZhou-Joe — APPROVED** [review]

**Review by ruanyl — APPROVED** [review]

**Pull request merged**